src/video.rs

```diff
@@ -115,34 +115,61 @@ fn pretty_spinner_style() -> ProgressStyle {
         .progress_chars(PROGRESS_CHARS)
 }
 
+type VideoCompareMutex<E, F> = Arc<Mutex<VideoCompare<E, F>>>;
+
+struct VideoCompare<E: Decoder, F: Decoder> {
+    current_frame: usize,
+    next_frame: usize,
+    source: E,
+    distorted: F,
+}
+
 fn calc_score<S: Pixel, D: Pixel, E: Decoder, F: Decoder>(
-    mtx: &Mutex<(usize, (E, F))>,
+    mtx: &VideoCompareMutex<E, F>,
     src_yuvcfg: &YuvConfig,
     dst_yuvcfg: &YuvConfig,
     inc: usize,
+    end_frame: Option<usize>,
     verbose: bool,
 ) -> Option<(usize, f64)> {
     let (frame_idx, (src_frame, dst_frame)) = {
         let mut guard = mtx.lock().unwrap();
-        let curr_frame = guard.0;
+        let mut curr_frame = guard.current_frame;
+        // We passed this as start frame.
+        // However, we are going to use it to store the next frame we should compute.
+        let mut next_frame = guard.next_frame;
+
+        //println!("{curr_frame} < {next_frame}");
+
+        while curr_frame < next_frame {
+            let _src_frame = guard.source.read_video_frame::<S>();
+            let _dst_frame = guard.distorted.read_video_frame::<D>();
+            if _src_frame.is_none() || _dst_frame.is_none() {
+                break;
+            }
+            if verbose {
+                println!("Frame {}: skip", curr_frame);
+            }
+            curr_frame += 1;
+        }
 
-        let src_frame = guard.1 .0.read_video_frame::<S>();
-        let dst_frame = guard.1 .1.read_video_frame::<D>();
+        next_frame = curr_frame + inc;
 
-        if let (Some(sf), Some(df)) = (src_frame, dst_frame) {
-            // skip remaining frames in increment size
-            for ii in 1..inc {
-                let _src_frame = guard.1 .0.read_video_frame::<S>();
-                let _dst_frame = guard.1 .1.read_video_frame::<D>();
-                if _src_frame.is_none() || _dst_frame.is_none() {
-                    break;
-                }
-                if verbose {
-                    println!("Frame {}: skip", curr_frame + ii);
-                }
+        if let Some(end_frame) = end_frame {
```

Comment:
Is this correct? Scenario: Current frame = 10, next frame = 15, increment = 5, end frame = 18

`while curr_frame < next_frame { skip_frames(); }` will increment `curr_frame` to 15.
`next_frame = curr_frame + inc;` will set `next_frame` to 20.

`if let Some(end_frame) = end_frame` is true, inner `end_frame` is 18. `next_frame` is now 20 so greater than `end_frame`, thus returning `None`. But logically, we should've still done the comparison since curr_frame is only 15 right?

Reply:
On further thought, I think this check should be done before skipping frames. Not 100% sure at a glance.

Reply:
I've readjusted the logic now. This check is done before skipping frames, this avoids another skip and read cycle before exiting.